### ISSUE TITLE
empire ID sequence wrongly mattering for success of competing colonization (and invasion) orders

### ISSUE THREAD

**Issue by Dilvish-fo** [freeorion/freeorion#1571]
In general, I think we've had an unstated (so far as I have noticed) presumption/rule that your particular empire ID, at least regarding where it fits into the overall sequence of empire IDs present in a game, should not actually affect any results of your actions; I think we don't quite succeed with this.  I am hoping someone can confirm my reading of the code and/or ingame observations.

**Regarding colonizing/outposting**,  it appears that the [ServerApp colonization code intends that if two empires try to colonize the same planet on the same turn, they both should fail.](https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2216-L2218)  

However, it appears to me that because [Orders are executed empire by empire implicitly in sequence from low empireID to high empireID,](https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2702-L2710) 

coupled with [an exclusion of parallel colonization attempts in the ColonizationOrder code,](https://github.com/freeorion/freeorion/blob/master/util/Order.cpp#L553-L555) it appears to me that the result is that in the face of competing colonization orders the empire with the lower empireID will always succeed and the empire with the higher empireID will always fail.   I can't recall having clearly observed this in game, though it would be hard to be sure of in a normal single player game.    For single player games this works out as an advantage for the human player, which some might think good and others bad, but certainly for multiplayer games it is a bad handling.

I think that the proper handling would be to remove the competition exclusion from the ColonizationOrder code, and let the ServerApp code block both competing colonization attempts.

Regarding Invasions, I don't see any code directed towards specifically dealing with the case of empires attempting competing invasions of the same planet.  The result, which I have observed ingame, is that the empire with the lower ID will invade first, and then the empire with the higher ID will invade, typically capturing the planet from the first empire (all depends on the troop numbers involved, but since the first empires troops have been depleted by the original defenders, it is generally the second empire that wins).  Again, debatable result for a single player game, but definitely bad for multiplayer.

I think that the best handling here would be to have some code akin to that for colonization, to figure out if there are competing invasion attempts.  We could then possibly block both attempts as with colonization, although personally think it would be a little better to simply have a better resolution handling for this case where the defenders get split up against the invaders.  One possible way would be to rank the empires/natives involved by number of troops, then begin a cycle: (a) take the lowest number of troops, (b) divide it by the number of remaining empires still having troops involved to get an allocation against them (could be weighted by their troop levels), (c) zero out that lowest troop number, (d) subtract the allocation from the remaining empires, (e) if any of the groups drop to zero troops remove them from contention, (f) rinse and repeat.

**Comment by geoffthemedio** [comment]
Re: colonization, removing the double-colonization in the order makes sense. It's intended to stop a single empire from colonizing the same planet twice, but seemingly could end up preventing multiple empires from colonizing.

Re: invasion, parallel invasions are supposed to be resolved all together. Several empires should be able to drop troops onto the same planet, then all empires' troops will battle once, and whoever has the most should win. Order execution just marks ships as "invading" (or "colonizing") and the actual dropping of troops (or colonists) happens later in HandleInvasion (or HandleColonization).

For invasion specifically, it shouldn't care if there are more than two empires with troops on a planet when the ground combat is resolved, or what order their orders were executed in:
https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2414
https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2297

What might need fixing is how invasion order timing during a turn interacts with diplomatic status changes... eg. can you order an invasion, make peace with the target, and still have the invasion go through? It should probably cancel all relevant invasions if peace breaks out mid-turn. There probably also needs to be a more-clearly enforced / demarkated blackout period during turn resolution during which diplomatic changes can't take effect... eg. after all empires have submitted turn orders, until new turn orders are sent out.

**Comment by Vezzra** [comment]
*Colonization:* I agree, if several empires try to colonize a planet simultaneously, all these attempts should fail (sidenote: regardless of diplomatic relations).

*Invasion*: Agree with multiple simultaneous invasion attempts by different empires should be possible, and the empire with the highest number of troops should win. The question is, how much troops the winner of a multi-empire planetary battle has left. A very simple resolution would be to just substract the troop strength of the second strongest empire from the troop strength of the victor. IMO we don't need anything more complicated.

However, there is one thing wrt to multiple simultaneous invasions that hasn't been mentioned at all yet, but needs to be accounted for: what happens if *allied* empires invade a planet simultaneously? Even better: what happens if there are several factions of allied empires invading a planet simultaneously? As exotic and rare that case might be, it's possible and going to happen, especially in multiplayer games.

Currently I expect the result to be seriously buggy, as the allies will fight each other (certainly not the desired behaviour).

My suggestion would be to determine allied empires prior to combat resolution, add the troop strength of allied empires and set the troop strength of the strongest member of an alliance to the combined troop strength. Which basically means the empire with the highest troop strength takes command of all the allied troops, and also takes over the planet if the alliance wins.

**Comment by geoffthemedio** [comment]
I meant to mention the allied invasions case Vezzra describes, but forgot while in the shower...

**Comment by Dilvish-fo** [comment]
Hmm, yes, Geoff, it appears I didn't look closely enough at the invasion code, and the current handling appears to be the leader-runner_up approach that Vezzra describes (I really thought I had observed it working differently ingame, but it appears that was most likely observer-error).  Also, with respect to alliances engaging in parallel invasion attempts, I think that Vezzra's proposal sounds sensible, and looks to me like it would be not too hard to work into the current invasion handling code.

As for colonization, just a slight bit more thought that still leaves me agreeing with Geoff that we should remove the double colonization check from the order:  I suppose one possibility could be that rather than entirely removing the parallel-colonization check we could simply make it more specific to parallel colonization by the same empire; however, I think that with regards to human players our UI already does an independent check that would catch that (and the UI really is the place it should be caught for human players rather than having a colonization order simply fail on execution), and likewise for AIs we should really just make sure the AI doesn't try that rather than letting it try but then having the order fail.

**Comment by geoffthemedio** [comment]
@Oberlus do you recall or can you check if this has been addressed? perhaps the check here https://github.com/freeorion/freeorion/blob/master/util/Order.cpp#L817 should be removed?

**Comment by Oberlus** [comment]
I've created a tiny MP game for this. I can confirm in current master 3523697 this is not addressed: player Oberlus (host, first empire) gets the colony, player Oberlus2 doesn't. (As expected, the order in which the players issue the colonization orders has no effect.)

Then I've commented out the check in https://github.com/freeorion/freeorion/blob/master/util/Order.cpp#L817-L820, reloaded the save, reissued the colonization orders, and the planet was not colonized.

I made the PR #4131
